core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedBinaryRange.java

```diff
@@ -0,0 +1,170 @@
+import org.elasticsearch.search.aggregations.Aggregation;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.ParsedMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.bucket.range.ip.IpRangeAggregationBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+public class ParsedBinaryRange extends ParsedMultiBucketAggregation<ParsedBinaryRange.ParsedBucket> implements Range {
+
+    @Override
+    public String getType() {
+        return IpRangeAggregationBuilder.NAME;
+    }
+
+    @Override
+    public List<? extends Range.Bucket> getBuckets() {
+        return buckets;
+    }
+
+    private static ObjectParser<ParsedBinaryRange, Void> PARSER =
+            new ObjectParser<>(ParsedBinaryRange.class.getSimpleName(), true, ParsedBinaryRange::new);
+    static {
+        declareMultiBucketAggregationFields(PARSER,
+                parser -> ParsedBucket.fromXContent(parser, false),
+                parser -> ParsedBucket.fromXContent(parser, true));
+    }
+
+    public static ParsedBinaryRange fromXContent(XContentParser parser, String name) throws IOException {
+        ParsedBinaryRange aggregation = PARSER.parse(parser, null);
+        aggregation.setName(name);
+        return aggregation;
+    }
+
+    public static class ParsedBucket extends ParsedMultiBucketAggregation.ParsedBucket implements Range.Bucket {
+
+        private String key;
+        private BytesRef from;
+        private BytesRef to;
+
+        @Override
+        public Object getKey() {
+            return key;
+        }
+
+        @Override
+        public String getKeyAsString() {
+            return key;
+        }
+
+        @Override
+        public Object getFrom() {
+            return getFromAsString();
+        }
+
+        @Override
+        public String getFromAsString() {
+            return bytesRefToString(from);
+        }
+
+        @Override
+        public Object getTo() {
+            return getToAsString();
+        }
+
+        @Override
+        public String getToAsString() {
+            return bytesRefToString(to);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            String key = this.key;
+            if (isKeyed()) {
+                if (key == null) {
+                    StringBuilder keyBuilder = new StringBuilder();
+                    keyBuilder.append(from == null ? "*" : getFromAsString());
+                    keyBuilder.append('-');
+                    keyBuilder.append(to == null ? "*" : getToAsString());
+                    key = keyBuilder.toString();
+                }
+                builder.startObject(key);
+            } else {
+                builder.startObject();
+                if (key != null) {
```

Comment:
Yes, the key can be null. I didn't handle this case correctly so I changed InternalBinaryRangeTests so that it randomly generates a bucket with a null key. I also changed the bucket parsing so that it handles keyed buckets with null keys.